src/lib/support/jsontlv/TlvJson.cpp

```diff
@@ -0,0 +1,235 @@
+        json[keyBuf] = val;
+    }
+    else if (keyContext.keyType == KeyContext::kStructField)
+    {
+        snprintf(keyBuf, sizeof(keyBuf), "%" PRIu32, keyContext.key);
+        json[keyBuf] = val;
+    }
+    else
+    {
+        json[keyContext.key] = val;
+    }
+}
+
+std::string JsonToString(Json::Value & json)
+{
+    Json::StyledWriter writer;
+    return writer.write(json);
+}
+
+CHIP_ERROR TlvToJson(TLV::TLVReader & reader, KeyContext context, Json::Value & parent)
+{
+    bool isStruct = false;
+
+    switch (reader.GetType())
+    {
+    case TLV::kTLVType_UnsignedInteger: {
+        uint64_t v;
+        ReturnErrorOnFailure(reader.Get(v));
+        InsertKeyValue(parent, context, v);
+        break;
+    }
+
+    case TLV::kTLVType_SignedInteger: {
+        int64_t v;
+        ReturnErrorOnFailure(reader.Get(v));
+        InsertKeyValue(parent, context, v);
+        break;
+    }
+
+    case TLV::kTLVType_Boolean: {
+        bool v;
+        ReturnErrorOnFailure(reader.Get(v));
+        InsertKeyValue(parent, context, v);
+        break;
+    }
+
+    case TLV::kTLVType_FloatingPointNumber: {
+        double v;
+        ReturnErrorOnFailure(reader.Get(v));
+        InsertKeyValue(parent, context, v);
+        break;
+    }
+
+    case TLV::kTLVType_ByteString: {
+        ByteSpan span;
+
+        ReturnErrorOnFailure(reader.Get(span));
+        VerifyOrReturnError(span.size() < kMaxStringLen, CHIP_ERROR_INVALID_TLV_ELEMENT);
+
+        Platform::ScopedMemoryBuffer<char> byteString;
+        byteString.Alloc(BASE64_ENCODED_LEN(span.size()) + 1);
+        VerifyOrReturnError(byteString.Get() != nullptr, CHIP_ERROR_NO_MEMORY);
+
+        auto encodedLen              = Base64Encode(span.data(), span.size(), byteString.Get());
+        byteString.Get()[encodedLen] = '\0';
+
+        InsertKeyValue(parent, context, byteString.Get());
+        break;
+    }
+
+    case TLV::kTLVType_UTF8String: {
+        CharSpan span;
+
+        ReturnErrorOnFailure(reader.Get(span));
+        VerifyOrReturnError(span.size() < kMaxStringLen, CHIP_ERROR_INVALID_TLV_ELEMENT);
+
+        Platform::ScopedMemoryString charString(span.data(), span.size());
+        InsertKeyValue(parent, context, charString.Get());
+        break;
+    }
+
+    case TLV::kTLVType_Null: {
+        InsertKeyValue(parent, context, Json::Value());
+        break;
+    }
+
+    case TLV::kTLVType_Structure:
+        isStruct = true;
+
+    case TLV::kTLVType_Array: {
+        TLV::TLVType containerType;
+
+        ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+        CHIP_ERROR err;
+        Json::Value value;
+        size_t listIndex = 0;
+
+        while ((err = reader.Next()) == CHIP_NO_ERROR)
+        {
+            if (isStruct)
+            {
+                KeyContext context2(static_cast<chip::FieldId>(TLV::TagNumFromTag(reader.GetTag())));
+                VerifyOrReturnError(TLV::IsContextTag(reader.GetTag()), CHIP_ERROR_INVALID_TLV_TAG);
```

Comment:
We should not be calling TagNumFromTag _before_ checking the tag type.

But also, what about profile tags here?  Why are we forcing context tags?

Reply:
> We should not be calling TagNumFromTag before checking the tag type.

Will fix.

> But also, what about profile tags here? Why are we forcing context tags?

There isn't a technical limitation here (a fully qualified profile tag = 8 octets, so should fit within a 64-bit integer in JSON), but I wanted to hew to the fact that this is meant to only be used with cluster payloads, within which such tags do not appear. A generalized TLV -> JSON converter needs more provisions for the full generality of TLV, which I didn't want to take on (nor felt necessary) for this initial cut.

Reply:
This imposed limitation is described in the comment body for `TlvToJson` (see [here](https://github.com/project-chip/connectedhomeip/pull/15440/files#diff-c8ffeb400cb088480209149c6a4472cd3b79872473ae3397ccb724b2a0e2dbaaR28)).